From 619e046807d59ce21bc85c5e06f74b399b11a8e3 Mon Sep 17 00:00:00 2001
From: Damien Hourtoulle <damien.hourtoulle@gmail.com>
Date: Tue, 14 Apr 2026 15:31:33 +0000
Subject: [PATCH] pahole: Fix discarded-qualifiers for strchr/strstr.

Glibc 2.43 added C23 const-preserving overloads :
https://sourceware.org/glibc/wiki/Release/2.43.

For the function prototype__new, fix local variable declaration to use
the correct const char* type instead of char*, removing the need to
discard the const qualifier.

Also fix a pre-existing bug in type__find_type_enum: when iterating
over a '+'-concatenated list of enum names, after advancing to the next
segment the new separator was not nullified, causing "second+third" to
be passed to cu__find_enumeration_by_name instead of "second".  Add a
null check before writing to sep_mutable to avoid a null dereference
when no further '+' is present.

Signed-off-by: Damien Hourtoulle <damien.hourtoulle@gmail.com>
Signed-off-by: Alan Maguire <alan.maguire@oracle.com>
Link: https://lore.kernel.org/dwarves/20260311140938.81459-2-d.hourtoulle@geoide.fr/
Signed-off-by: Vamsi Krishna Brahmajosyula <vamsi-krishna.brahmajosyula@broadcom.com>
---
 lib/bpf/src/libbpf.c |  2 +-
 pahole.c             | 35 ++++++++++++++++++-----------------
 2 files changed, 19 insertions(+), 18 deletions(-)

diff --git a/lib/bpf/src/libbpf.c b/lib/bpf/src/libbpf.c
index 0781fae..26a9780 100644
--- a/lib/bpf/src/libbpf.c
+++ b/lib/bpf/src/libbpf.c
@@ -11420,7 +11420,7 @@ static int resolve_full_path(const char *file, char *result, size_t result_sz)
 		if (!search_paths[i])
 			continue;
 		for (s = search_paths[i]; s != NULL; s = strchr(s, ':')) {
-			char *next_path;
+			const char *next_path;
 			int seg_len;
 
 			if (s[0] == ':')
diff --git a/pahole.c b/pahole.c
index e87d9a4..879571f 100644
--- a/pahole.c
+++ b/pahole.c
@@ -2263,7 +2263,7 @@ static int64_t type_instance__int_value(struct type_instance *instance, const ch
 	int byte_offset = 0;
 
 	if (!member) {
-		char *sep = strchr(member_name_orig, '.');
+		const char *sep = strchr(member_name_orig, '.');
 
 		if (!sep)
 			return -1;
@@ -2276,8 +2276,8 @@ static int64_t type_instance__int_value(struct type_instance *instance, const ch
 		char *member_name = member_name_alloc;
 		struct type *type = instance->type;
 
-		sep = member_name_alloc + (sep - member_name_orig);
-		*sep = 0;
+		char *sep_mutable = member_name_alloc + (sep - member_name_orig);  // sep mutable for the copy
+		*sep_mutable = 0;
 
 		while (1) {
 			member = type__find_member_by_name(type, member_name);
@@ -2290,9 +2290,9 @@ out_free_member_name:
 			type = tag__type(cu__type(cu, member->tag.type));
 			if (type == NULL)
 				goto out_free_member_name;
-			member_name = sep + 1;
-			sep = strchr(member_name, '.');
-			if (!sep)
+			member_name = sep_mutable + 1;
+			sep_mutable = strchr(member_name, '.');
+			if (!sep_mutable)
 				break;
 
 		}
@@ -2722,7 +2722,7 @@ static struct prototype *prototype__new(const char *expression)
 
 	strcpy(prototype->name, expression);
 
-	const char *name = prototype->name;
+	char *name = prototype->name;
 
 	prototype->nr_args = 0;
 
@@ -2736,10 +2736,9 @@ static struct prototype *prototype__new(const char *expression)
 	if (args_close == NULL)
 		goto out_no_closing_parens;
 
+	*args_open++ = *args_close = '\0';
 	char *args = args_open;
 
-	*args++ = *args_close = '\0';
-
 	while (isspace(*args))
 		++args;
 
@@ -2886,7 +2885,7 @@ static int type__find_type_enum(struct type *type, struct cu *cu, const char *ty
 		return type__add_type_enum(type, te, cu);
 
 	// Now look at a 'virtual enum', i.e. the concatenation of multiple enums
-	char *sep = strchr(type_enum, '+');
+	const char *sep = strchr(type_enum, '+');
 
 	if (!sep)
 		return -1;
@@ -2898,13 +2897,13 @@ static int type__find_type_enum(struct type *type, struct cu *cu, const char *ty
 
 	int ret = -1;
 
-	sep = type_enums + (sep - type_enum);
+	char *sep_mutable = type_enums + (sep - type_enum);
+	char *cur = type_enums;
 
-	type_enum = type_enums;
-	*sep = '\0';
+	*sep_mutable = '\0';
 
 	while (1) {
-		te = cu__find_enumeration_by_name(cu, type_enum, NULL);
+		te = cu__find_enumeration_by_name(cu, cur, NULL);
 
 		if (!te)
 			goto out;
@@ -2913,10 +2912,12 @@ static int type__find_type_enum(struct type *type, struct cu *cu, const char *ty
 		if (ret)
 			goto out;
 
-		if (sep == NULL)
+		if (sep_mutable == NULL)
 			break;
-		type_enum = sep + 1;
-		sep = strchr(type_enum, '+');
+		cur = sep_mutable + 1;
+		sep_mutable = strchr(cur, '+');
+		if (sep_mutable)
+			*sep_mutable = '\0';
 	}
 
 	ret = 0;
-- 
2.43.7

